runtime/src/iree-amd-aie/driver/xrt-lite/device.cc

```diff
@@ -0,0 +1,296 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree-amd-aie/driver/xrt-lite/shim/linux/kmq/device.h"
+
+#include "iree-amd-aie/driver/xrt-lite/allocator.h"
+#include "iree-amd-aie/driver/xrt-lite/api.h"
+#include "iree-amd-aie/driver/xrt-lite/direct_command_buffer.h"
+#include "iree-amd-aie/driver/xrt-lite/nop_executable_cache.h"
+#include "iree-amd-aie/driver/xrt-lite/nop_semaphore.h"
+#include "iree-amd-aie/driver/xrt-lite/util.h"
+#include "iree/hal/utils/deferred_command_buffer.h"
+#include "iree/hal/utils/deferred_work_queue.h"
+
+#define ARENA_BLOCK_SIZE (32 * 1024)
+
+namespace {
+extern const iree_hal_device_vtable_t iree_hal_xrt_lite_device_vtable;
+}
+
+struct iree_hal_xrt_lite_device {
+  iree_hal_resource_t resource;
+  iree_string_view_t identifier;
+  iree_allocator_t host_allocator;
+  // TODO(max): not used because "device allocations" are performed through
+  // device
+  iree_hal_allocator_t* device_allocator;
+  // block pool used for command buffer allocations, uses a larger block size
+  // since command buffers can contain inlined data
+  iree_arena_block_pool_t block_pool;
+  shim_xdna::device* shim_device;
+
+  iree_hal_xrt_lite_device(const iree_hal_xrt_lite_device_options* options,
+                           iree_allocator_t host_allocator) {
+    IREE_ASSERT_ARGUMENT(options);
+    IREE_TRACE_ZONE_BEGIN(z0);
+
+    iree_hal_resource_initialize(&iree_hal_xrt_lite_device_vtable, &resource);
+    this->host_allocator = host_allocator;
+    shim_device = new shim_xdna::device;
+
+    iree_status_t status = iree_hal_xrt_lite_allocator_create(
+        host_allocator, shim_device, &device_allocator);
+    IREE_ASSERT(iree_status_is_ok(status));
+    iree_arena_block_pool_initialize(ARENA_BLOCK_SIZE, host_allocator,
+                                     &block_pool);
+
+    IREE_TRACE_ZONE_END(z0);
+  }
+};
+
+static iree_status_t iree_hal_xrt_lite_device_create_executable_cache(
+    iree_hal_device_t* base_value, iree_string_view_t identifier,
+    iree_loop_t loop, iree_hal_executable_cache_t** out_executable_cache) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_device* device =
```

Comment:
For type safety we often use a wrapper for these types of casts in iree. 
e.g.:
```

static iree_hal_hip_device_t* iree_hal_hip_device_cast(
    iree_hal_device_t* base_value) {
  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_hip_device_vtable);
  return (iree_hal_hip_device_t*)base_value;
}
```

Reply:
yea ofc i saw this everywhere but elided it assuming it was an artifact all the other HALs being written in C. but now investigating `reinterpret_cast` I see that it is indeed not type safe. so i'll change this back.